### PR TITLE
Docker for Open3D

### PR DESCRIPTION
**Pull request by takanokage**
Headless rendering on ubuntu:16.04 with Xvfb, x11vnc, lxde and bind mounting.

### CODE REVIEW

**Comment by takanokage** [comment]
please hold the merge, I need to make some tweaks.

File: utilities/docker/ubuntu-xvfb/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,67 @@
+# MAINTAINER Dan Petre dan.petre@intel.com
```
Comment (syncle) [inline utilities/docker/ubuntu-xvfb/Dockerfile]: We don't specify contributor's name for each code. Please use the following.

```
# Open3D: www.open3d.org
# The MIT License (MIT)
# See license file or visit www.open3d.org for details
```

Github will eventually track contributors line by line.

File: utilities/docker/ubuntu-xvfb/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+FROM ubuntu:16.04
+
+ENV http_proxy "http://proxy-chain.intel.com:911"
+ENV HTTP_PROXY "http://proxy-chain.intel.com:911"
+ENV https_proxy "http://proxy-chain.intel.com:912"
+ENV HTTPS_PROXY "http://proxy-chain.intel.com:912"
+ENV ftp_proxy "http://proxy-chain.intel.com:911"
+ENV socks_proxy "http://proxy-us.intel.com:1080"
+ENV no_proxy "intel.com;.intel.com;localhost;127.0.0.1"
+ENV ALL_PROXY "socks5://proxy-us.intel.com"
```
Comment (syncle) [inline utilities/docker/ubuntu-xvfb/Dockerfile]: Please remove proxy settings.

File: utilities/docker/ubuntu-xvfb/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    build-essential \
+    cmake \
+    git \
+    gdb \
+    # libeigen3-dev \
+    # libgl1-mesa-dev \
+    # libgl1-mesa-glx \
+    # libglew-dev \
+    # libglfw3-dev \
+    libglu1-mesa-dev \
```
Comment (syncle) [inline utilities/docker/ubuntu-xvfb/Dockerfile]: Please remove commented lines.

File: utilities/docker/ubuntu-xvfb/setup/headless_sample.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    vis.add_geometry(pcd)
+    vis.get_render_option().load_from_json("../../TestData/renderoption.json")
+    vis.register_animation_callback(move_forward)
+    vis.run()
+    vis.destroy_window()
+
+if __name__ == "__main__":
+    pcd = read_point_cloud("../../TestData/fragment.ply")
+
+    print("5. Customized visualization playing a camera trajectory")
```
Comment (syncle) [inline utilities/docker/ubuntu-xvfb/setup/headless_sample.py]: Remove ``5.``

**Comment by syncle** [comment]
We had decided to use `util` for the new folder name, not `utilities`.

**Comment by takanokage** [comment]
ok, let me change that in a second.

**Comment by syncle** [comment]
Thanks @takanokage. Can you also make a new PR to submit a document for this? We will need instructions to use this. You can work on ``docs/advanced/`` and make ``headless_rendering.rst.``

**Comment by takanokage** [comment]
Yep, that was the plan from the beginning. For the purpose of this particular PR I added [brief instructions](https://github.com/IntelVCL/Open3D/issues/17#issuecomment-377375251) to #17.

**Comment by takanokage** [comment]
I have written the headless rendering documentation page. I'm sure there are things that I missed. Please try the headless rendering demo and let me know what you would like improved. Let me know if/how I can help; I'd be happy to walk you trough if/when you get stuck.

**Comment by syncle** [comment]
Awesome. Thanks @takanokage! I am looking this PR. I will post any comments regarding this.

**Comment by syncle** [comment]
I just tested this PR. I was able to save both depth and color images. However, the saved color images are just filled with black color. Depth images looks fine. @takanokage, can you check on this?

**Comment by takanokage** [comment]
will do. could you please give me more info? how did you obtain the color images?

**Comment by syncle** [comment]
I just tried [headless rendering from terminal](https://github.com/IntelVCL/Open3D/pull/262/files#diff-34798eda4e5a1310b132fd56d45a1653) and use the following scripts. 

```
$ cd <Open3D path>/utilities/docker/ubuntu-xvfb/tools
$ ./build.sh
$ ./attach.sh
$ ./headless_sample.sh
```

**Comment by takanokage** [comment]
I'm doing the same thing but it looks like I'm only getting depth images in /Open3D/build/lib/TestData/depth. Did you do anything else?

**Comment by syncle** [comment]
Nope. I did not tweak anything.

>  I'm only getting depth images in /Open3D/build/lib/TestData/depth

I think you are also having same trouble. My case saves color images in /Open3D/build/lib/TestData/color, but they are just black images.

**Comment by takanokage** [comment]
ok, I confirm the images in /Open3D/build/lib/TestData/image are black.
There seems to be a problem with `visualizer.capture_screen_float_buffer(...)`. What's your opinion? I'm still ramping up on the Open3D code so any pointers would help.

**Comment by takanokage** [comment]
Another observation: I'm getting similar behavior (black images) when operating Open3D over VNC.

**Comment by takanokage** [comment]
`capture_screen_image` works.
`capture_depth_image` doesn't work (saves black images).

**Comment by qianyizh** [comment]
I thought we have discussed this in #17  and have decided to use only glfw3.3 dev branch for headless rendering?

**Comment by syncle** [comment]
True. I prefer GLFW3.3 dev. Dan tried this option before and he decide to keep using GLFW3.2 due to technical reason. This PR still based on GLFW 3.2. However, I believe this issue should be addressed with GLFW3.3 dev.

**Comment by takanokage** [comment]
well, we've discussed several topics related to #17.

a. Jaesik worked with glfw3.3 in order to get to headless rendering. I didn't try that path and don't know if it works.
b. I worked with Docker in order to get to headless rendering. glfw3.3 crashes in the Docker setup but glfw3.2 doesn't.

**Comment by takanokage** [comment]
I'd also like to point out that rendering works just fine as confirmed by VNC into the Docker container.

**Comment by takanokage** [comment]
I propose we break this down into two parts:
1. the current PR
2. another PR for the limitations/issues of the current headless solution.

Advantages:
- people get access to some headless functionality, they get to try it out and maybe even contribute to it.
- the current PR doesn't get delayed for any additional issue we find with the current solution

If we do this we need to list the current known limitations in a "release" doc.

**Comment by takanokage** [comment]
that sounds great, good luck!

**Comment by syncle** [comment]
Good news! **glew2.1 + OSMesa18.0 + glfw3.3 dev works for headless rendering.**
I was able to generate rendering images from terminal via clean Ubuntu 16.04. I don't have any black screen images. I don't need to make a fake display. It just runs out-of-box. 

I will need to polish how I did: dependencies, modifying cmake tool chains, adding switch "headless rendering" for selective compilation. This switch will feed one of graphical library (OpenGL or OSMesa) for glew and glfw.

**Comment by takanokage** [comment]
closing until #273 is over.

**Pull request closed without merging**